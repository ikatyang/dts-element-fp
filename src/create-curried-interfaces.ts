import * as dts from 'dts-element';
import * as R from 'ramda';
import {
  get_interface_generic_name_default,
  get_interface_generic_return_name_default,
  get_interface_name_default,
  get_interface_parameter_name_default,
  get_interface_parameter_placeholder_name_default,
  get_placeholder_type_default,
  placeholder_default,
  selectable_default,
} from './utils/constants';
import {
  create_selectable_signatures,
  CreateSelectableSignaturesOptions,
} from './utils/create-selectable-signatures';
import { push_signatures } from './utils/push-signatures';
import { sort_signatures } from './utils/sort-signatures';

export interface CreateCurriedInterfacesOptions
  extends CreateSelectableSignaturesOptions {
  selectable?: boolean;
  placeholder?: boolean;
  get_placeholder_type?: () => string;
  get_interface_name?: (index: number) => string;
  get_generic_name?: (index: number) => string;
  get_generic_return_name?: () => string;
  get_parameter_name?: (index: number) => string;
  get_parameter_placeholder_name?: (index: number) => string;
}

export function create_curried_interfaces(
  max_curry_level: number,
  // istanbul ignore next
  options: CreateCurriedInterfacesOptions = {},
): dts.IInterfaceDeclaration[] {
  // istanbul ignore next
  const {
    selectable = selectable_default,
    placeholder = placeholder_default,
    get_placeholder_type = get_placeholder_type_default,
    get_interface_name = get_interface_name_default,
    get_generic_name = get_interface_generic_name_default,
    get_generic_return_name = get_interface_generic_return_name_default,
    get_parameter_name = get_interface_parameter_name_default,
    get_parameter_placeholder_name = get_interface_parameter_placeholder_name_default,
    get_selectable_kind_name,
    get_selectable_selector_name,
  } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const generics = R.range(0, max_curry_level).map((_, index) =>
    dts.create_generic_declaration({ name: get_generic_name(index) }),
  );

  const generic_return = dts.create_generic_declaration({
    name: get_generic_return_name(),
  });

  const parameters = generics.map((generic, index) =>
    dts.create_parameter_declaration({
      name: get_parameter_name(index),
      type: dts.create_general_type({ name: generic.name }),
    }),
  );

  const placeholders = parameters.map((_, index) =>
    dts.create_parameter_declaration({
      name: get_parameter_placeholder_name(index),
      type: dts.create_general_type({ name: get_placeholder_type() }),
    }),
  );

  const interfaces = generics.map((_, index) =>
    dts.create_interface_declaration({
      name: get_interface_name(index),
      generics: [...generics.slice(0, index + 1), generic_return],
      type: dts.create_object_type({ members: [] }),
    }),
  );

  interfaces.forEach((the_interface, index) => {
    push_signatures(
      is_placeholder,
      parameters.slice(0, index + 1),
      placeholder ? placeholders.slice(0, index + 1) : [],
      (used_parameters, return_parameters) => {
        the_interface.type!.members!.push(
          dts.create_object_member({
            owned: dts.create_function_declaration({
              name: undefined,
              type: dts.create_function_type({
                parameters: used_parameters,
                return:
                  return_parameters.length === 0
                    ? dts.create_general_type({ name: generic_return.name })
                    : dts.create_general_type({
                        name: interfaces[return_parameters.length - 1].name,
                        generics: [
                          ...return_parameters.map(
                            parameter => parameter.type!,
                          ),
                          dts.create_general_type({
                            name: generic_return.name,
                          }),
                        ],
                      }),
              }),
            }),
          }),
        );
      },
    );

    const object_type = interfaces[index].type!;
    object_type.members = sort_signatures(
      object_type.members as dts.IObjectMember[],
      { is_placeholder },
    );

    if (!selectable || object_type.members.length <= 1) {
      return;
    }

    object_type.members.splice(
      -1,
      0,
      ...create_selectable_signatures(
        object_type.members as dts.IObjectMember[],
        {
          get_selectable_kind_name,
          get_selectable_selector_name,
          is_placeholder,
        },
      ).reverse(),
    );
  });

  return interfaces;
}
