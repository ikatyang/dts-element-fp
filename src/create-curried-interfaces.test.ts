import * as dts from 'dts-element';
import { create_curried_interfaces } from './create-curried-interfaces';

it('should generate correctly without any options', () => {
  expect(emit_curried_interfaces()).toMatchSnapshot();
});

it('should generate correctly with selectable', () => {
  expect(
    emit_curried_interfaces({
      selectable: true,
    }),
  ).toMatchSnapshot();
});

it('should generate correctly with placeholder', () => {
  expect(
    emit_curried_interfaces({
      placeholder: true,
    }),
  ).toMatchSnapshot();
});

it('should generate correctly with every option', () => {
  expect(
    emit_curried_interfaces({
      selectable: true,
      placeholder: true,
    }),
  ).toMatchSnapshot();
});

function emit_curried_interfaces(
  options: { selectable?: boolean; placeholder?: boolean } = {},
) {
  const curried_interfaces = create_curried_interfaces(3, options);
  const top_level_element = dts.create_top_level_element({
    members: curried_interfaces,
  });
  return dts.emit(top_level_element);
}
