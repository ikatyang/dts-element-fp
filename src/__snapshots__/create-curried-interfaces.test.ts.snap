// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate correctly with every option 1`] = `
"interface CurriedFunction1<T1, R> {
    (v1: T1): R;
}
interface CurriedFunction2<T1, T2, R> {
    (v1: T1): CurriedFunction1<T2, R>;
    (_1: PH, v2: T2): CurriedFunction1<T1, R>;
    <$SEL extends \\"1\\">(): (v1: T1) => CurriedFunction1<T2, R>;
    <$SEL extends \\"01\\">(): (_1: PH, v2: T2) => CurriedFunction1<T1, R>;
    <$SEL extends \\"11\\">(): (v1: T1, v2: T2) => R;
    (v1: T1, v2: T2): R;
}
interface CurriedFunction3<T1, T2, T3, R> {
    (v1: T1): CurriedFunction2<T2, T3, R>;
    (_1: PH, v2: T2): CurriedFunction2<T1, T3, R>;
    (v1: T1, v2: T2): CurriedFunction1<T3, R>;
    (_1: PH, _2: PH, v3: T3): CurriedFunction2<T1, T2, R>;
    (_1: PH, v2: T2, v3: T3): CurriedFunction1<T1, R>;
    (v1: T1, _2: PH, v3: T3): CurriedFunction1<T2, R>;
    <$SEL extends \\"1\\">(): (v1: T1) => CurriedFunction2<T2, T3, R>;
    <$SEL extends \\"01\\">(): (_1: PH, v2: T2) => CurriedFunction2<T1, T3, R>;
    <$SEL extends \\"11\\">(): (v1: T1, v2: T2) => CurriedFunction1<T3, R>;
    <$SEL extends \\"001\\">(): (_1: PH, _2: PH, v3: T3) => CurriedFunction2<T1, T2, R>;
    <$SEL extends \\"011\\">(): (_1: PH, v2: T2, v3: T3) => CurriedFunction1<T1, R>;
    <$SEL extends \\"101\\">(): (v1: T1, _2: PH, v3: T3) => CurriedFunction1<T2, R>;
    <$SEL extends \\"111\\">(): (v1: T1, v2: T2, v3: T3) => R;
    (v1: T1, v2: T2, v3: T3): R;
}
"
`;

exports[`should generate correctly with placeholder 1`] = `
"interface CurriedFunction1<T1, R> {
    (v1: T1): R;
}
interface CurriedFunction2<T1, T2, R> {
    (v1: T1): CurriedFunction1<T2, R>;
    (_1: PH, v2: T2): CurriedFunction1<T1, R>;
    (v1: T1, v2: T2): R;
}
interface CurriedFunction3<T1, T2, T3, R> {
    (v1: T1): CurriedFunction2<T2, T3, R>;
    (_1: PH, v2: T2): CurriedFunction2<T1, T3, R>;
    (v1: T1, v2: T2): CurriedFunction1<T3, R>;
    (_1: PH, _2: PH, v3: T3): CurriedFunction2<T1, T2, R>;
    (_1: PH, v2: T2, v3: T3): CurriedFunction1<T1, R>;
    (v1: T1, _2: PH, v3: T3): CurriedFunction1<T2, R>;
    (v1: T1, v2: T2, v3: T3): R;
}
"
`;

exports[`should generate correctly with selectable 1`] = `
"interface CurriedFunction1<T1, R> {
    (v1: T1): R;
}
interface CurriedFunction2<T1, T2, R> {
    (v1: T1): CurriedFunction1<T2, R>;
    <$SEL extends \\"1\\">(): (v1: T1) => CurriedFunction1<T2, R>;
    <$SEL extends \\"11\\">(): (v1: T1, v2: T2) => R;
    (v1: T1, v2: T2): R;
}
interface CurriedFunction3<T1, T2, T3, R> {
    (v1: T1): CurriedFunction2<T2, T3, R>;
    (v1: T1, v2: T2): CurriedFunction1<T3, R>;
    <$SEL extends \\"1\\">(): (v1: T1) => CurriedFunction2<T2, T3, R>;
    <$SEL extends \\"11\\">(): (v1: T1, v2: T2) => CurriedFunction1<T3, R>;
    <$SEL extends \\"111\\">(): (v1: T1, v2: T2, v3: T3) => R;
    (v1: T1, v2: T2, v3: T3): R;
}
"
`;

exports[`should generate correctly without any options 1`] = `
"interface CurriedFunction1<T1, R> {
    (v1: T1): R;
}
interface CurriedFunction2<T1, T2, R> {
    (v1: T1): CurriedFunction1<T2, R>;
    (v1: T1, v2: T2): R;
}
interface CurriedFunction3<T1, T2, T3, R> {
    (v1: T1): CurriedFunction2<T2, T3, R>;
    (v1: T1, v2: T2): CurriedFunction1<T3, R>;
    (v1: T1, v2: T2, v3: T3): R;
}
"
`;
